File: processor/src/main/java/com/github/gfx/static_gson/ModelDefinition.java
package com.github.gfx.static_gson;

import com.github.gfx.static_gson.annotation.JsonSerializable;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.DeclaredType;
import javax.tools.Diagnostic;

public class ModelDefinition {

    public final TypeElement element;

    public final ClassName modelType;

    public final TypeRegistry typeRegistry = new TypeRegistry();

    private final StaticGsonContext context;

    private final List<FieldDefinition> fields;

    public ModelDefinition(StaticGsonContext context, TypeElement element) {
        this.context = context;
        this.element = element;

        modelType = ClassName.get(element);

        JsonSerializable annotation = element.getAnnotation(JsonSerializable.class);

        fields = new ArrayList<>();

        while (true) {
            fields.addAll(extractFields(annotation, element));
            if (element.getSuperclass().toString().equals(Object.class.getName())) {
                // reached the root
                break;
            }

            TypeElement superElement = context.getTypeElement(element.getSuperclass());

            if (superElement == null) {
                context.processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR,
                        String.format("Unable create static gson for %s. Perhaps this is a generic type which is not supported yet", element.getSuperclass().toString()));
                break;
            }

            element = superElement;
        }
    }

    private static List<FieldDefinition> extractFields(
            JsonSerializable config,
            TypeElement typeElement) {

        return typeElement.getEnclosedElements().stream()
                .filter(element -> element instanceof VariableElement)
                .map(element -> (VariableElement) element)
                .filter(element -> !element.getModifiers().contains(Modifier.TRANSIENT))
                .filter(element -> !element.getModifiers().contains(Modifier.STATIC))
                .map(element -> new FieldDefinition(config, element))
                .collect(Collectors.toList());
    }

    public Set<TypeName> getComplexTypes() {
        return fields.stream()
                .filter(field -> !field.isSimpleType())
                .map(FieldDefinition::getType)
                .collect(Collectors.toSet());
    }

    public List<FieldDefinition> getFields() {
        return fields;
    }
}

File: processor/src/main/java/com/github/gfx/static_gson/ModelDefinition.kt
package com.github.gfx.static_gson

import com.github.gfx.static_gson.annotation.JsonSerializable
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.TypeName
import org.jetbrains.kotlin.metadata.jvm.deserialization.JvmProtoBufUtil
import org.jetbrains.kotlin.serialization.deserialization.getName
import java.util.*
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.element.VariableElement
import javax.tools.Diagnostic

class ModelDefinition(private val context: StaticGsonContext, element: TypeElement) {

    private val METADATA_ANNOTATION_NAME = "Metadata"

    val element: TypeElement

    val modelType: ClassName

    val typeRegistry = TypeRegistry()

    private val fields: MutableList<FieldDefinition>

    val isKotlin: Boolean

    val complexTypes: Set<TypeName>
        get() = fields
                .filter { !it.isSimpleType }
                .map { it.type }
                .toSet()

    init {
        var element = element
        this.element = element

        modelType = ClassName.get(element)

        val annotation = element.getAnnotation(JsonSerializable::class.java)
        isKotlin = AnnotationHelper.hasAnnotationWithName(element, METADATA_ANNOTATION_NAME)

        if (isKotlin && !AnnotationHelper.hasPublicParameterlessConstructor(element)) {

            val metadata = kotlin.AnnotationHelper.getMetadata(element)!!

            val (nameResolver, classProto) = JvmProtoBufUtil.readClassDataFrom(metadata.d1, metadata.d2)

            if (classProto.constructorList.size > 1) {
                context.processingEnv.messager.printMessage(Diagnostic.Kind.ERROR,
                        String.format(
                                "Kotlin classes %s must have only 1 constructor",
                                element.toString()))
            }

            val hasAnyDefaultParameterSet = classProto.constructorList[0].valueParameterList.any { it.flags and 0b00000010 > 0 }

            if (hasAnyDefaultParameterSet) {
                val hasAllDefaultParameterSet = classProto.constructorList[0].valueParameterList.all { it.flags and 0b00000010 > 0 }
                if (!hasAllDefaultParameterSet) {
                    context.processingEnv.messager.printMessage(Diagnostic.Kind.ERROR, "All default parameters on $element must be set if some are set")
                    classProto.constructorList[0].valueParameterList
                            .filter { it.flags and 0b00000010 == 0 }
                            .forEach {
                                context.processingEnv.messager.printMessage(Diagnostic.Kind.ERROR, "Default parameters of \"$element.${nameResolver.getName(it.name).asString()}\" should be set.")
                            }

                }
            }
        }

        fields = ArrayList()

        while (true) {
            fields.addAll(extractFields(annotation, element, isKotlin, context))
            if (element.superclass.toString() == Any::class.java.name) {
                // reached the root
                break
            }

            val superElement = context.getTypeElement(element.superclass)

            if (superElement == null) {
                context.processingEnv.messager.printMessage(Diagnostic.Kind.ERROR,
                        String.format(
                                "Unable create static gson for %s. Perhaps this is a generic type which is not supported yet",
                                element.superclass.toString()))
                break
            }

            element = superElement
        }
    }

    fun getFields(): List<FieldDefinition> {
        return fields
    }


    private fun extractFields(
            config: JsonSerializable,
            typeElement: TypeElement,
            isKotlin: Boolean,
            context: StaticGsonContext): List<FieldDefinition> {

        return typeElement.enclosedElements
                .filter { element -> element is VariableElement }
                .map { element -> element as VariableElement }
                .filter { element -> !element.modifiers.contains(Modifier.TRANSIENT) }
                .filter { element -> !element.modifiers.contains(Modifier.STATIC) }
                .map { element ->
                    FieldDefinition(config, element, isKotlin,
                            AnnotationHelper.hasDeclaredDefault(typeElement, element))
                }

    }
}
